### ISSUE TITLE
정규화

### ISSUE THREAD

**Issue by 98hyun** [Sejong-Kaggle-Challengers/MAIN#2]
<h2>아파트 경매가격 예측 경진대회</h2>

회귀

<h4>정규화</h4>

rmse 평가지표. 루트평균제곱오차  
이번 목적함수는 회귀에 맞는 l2 loss 로 했다.   
<h4>l1 정규화, l2 정규화.에 대해서</h4>

norm : 벡터의 크기를 측정하는. 계산하는 방법이다. 아래 참고.   
p=1 일 때, l1 즉, 맨해튼거리이고 p=2일때 l2 유클리드 거리다. 이거는 외우기 싶게 한 뜻이고. 아래 참고.

식 : <img src="https://render.githubusercontent.com/render/math?math=||x||_p = (\sum_{i=1}^n |x_i|^p)^{1/p}"/>

<h4>ridge lasso</h4>

train대로만 학습이되면 overfitting의 위험이 있다. 그래서 손실함수(Cost)에 규제를 준다.    
가중치의 합으로 규제를 주면 l1 규제. 가중치의 제곱의 합으로 규제를 주면 l2 규제.   
정규화와 비슷하다.  

l1 규제를 사용하는것이 lasso. l2 규제를 사용한것이 ridge. 둘다 사용하는것이 elastic net.  

<h4>ridgecv lassocv</h4>

<h4>손실함수</h4>

실제값과 예측값의 차이를 손실함수의 값으로 정의한다.    
rmse는 손실함수를 적게 해야하므로 손실함수를 최소화 하는쪽으로 가중치를 업데이트한다.
그래서 pytorch를 이용해 구현했다.

<h4>pytorch</h4>  

[공부한 곳](https://wikidocs.net/book/2788)  
autograd로 미분값 자동 계산. opimizer로 w,b 업데이트.

<h4>참고</h4>

l1,l2 정규화 - [참고](https://light-tree.tistory.com/125)

맨해튼거리, 유클리드 거리 - [맨해튼](https://needjarvis.tistory.com/455), [유클리드](https://needjarvis.tistory.com/454)

**Comment by 98hyun** [comment]
```python
# 시도
from sklearn.linear_model import RidgeCV
from sklearn.model_selection import cross_val_score

model=RidgeCV()

scores=cross_val_score(model,x_train,y_train,cv=5)
scores 
# array([-2.31817001e+13, -4.04998287e+13, -9.75433588e+13, -1.07703338e+09,
#  -4.04759051e+14])

# 다른 방법 
# https://www.theteams.kr/teams/891/post/69522

from sklearn.model_selection import KFold
kf = KFold(5)
kf.get_n_splits()
scv = []
for train_idx, test_idx in kf.split(x_train):
    model.fit(x_train.iloc[train_idx],y_train.iloc[train_idx])
    s = model.score(x_train.iloc[test_idx],y_train.iloc[test_idx])
    scv.append(s) 

scv
# [-23181700143163.44,
#  -40499828712006.62,
#  -97543358786775.11,
#  -1077033376.2604933,
#  -404759051260844.9]
```

이번에 baseline을 pycaret으로 돌렸는데 제일 높은 점수로 ridge 뜨더라구요. 코드는 제 repo에 있습니다. 그런데 똑같이 해봤는데 같은 점수가 안나와요. 뭐가 문제인지 모르겠습니다. 혹시라도 아시면 공유부탁드릴게요.
